lib/active_model/serializer/adapter.rb

```diff
@@ -94,6 +98,38 @@ def include_meta(json)
         json[meta_key] = meta if meta
         json
       end
+
+      def key_format
+        serializer.key_format || default_key_format
+      end
+
+      def format_keys(formattable)
+        return formattable if !formattable.respond_to? :each_with_object ||
+                              key_format == :unaltered
+
+        if formattable.respond_to? :keys
+          formattable.each_with_object({}) do |(key, value), formatted_hash|
+            formatted_hash[format_key(key)] = format_keys(value)
```

Comment:
idle question sparked by recursion: might we want to cache key conversions (like tail call optimization)? I imagine many of them repeat.

Also, the JSON library has a `recurse_proc` method, which might be useful for doing the key formatting in`def to_json` rather than `def serializable_hash`